README
NAME
    cpanf - Query and install CPAN modules realtime from the fresh mirror

SYNOPSIS
      # install Module::Name from the fresh mirror
      cpanf Module::Name

      # list the recent uploads
      cpanf -l

      # search the latest uploads for "Catalyst"
      cpanf -l Catalyst

DESCRIPTION
    `cpanf' is a command line tool to find, index and download CPAN modules
    from fresh mirrors. It would be useful if you want to install modules
    before it's mirrored to the mirrors near you which usually takes a day
    or two.

    This tool utilizes FriendFeed CPAN relatime bot
    (http://friendfeed.com/cpan) to search and query the updates and CPAN
    Testers mirror (http://cpan.cpantesters.org/) to download modules.

AUTHOR
    Tatsuhiko Miyagawa <miyagawa@bulknews.net>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    App::CPAN::Fresh http://friendfeed.com/cpan



Changes
Revision history for Perl extension App::CPAN::Fresh

{{$NEXT}}

0.12  2015-04-17 17:07:08 CEST
        - Add deprecation message

0.11  Fri Nov  9 09:12:21 PST 2012
        - Support v-string versions

0.10  Wed Feb 16 10:15:17 PST 2011
        - Support -TRIAL releases

0.09  Tue Jun 22 17:37:35 EDT 2010
        - Fixed -h docs 

0.08  Fri Feb 19 15:19:26 PST 2010
        - Added --minus|-m option to use cpanminus http://bit.ly/cpanm

0.07  Fri Sep 11 16:15:25 JST 2009
        - Added -d option to install devel version of the module.

0.06  Wed Sep  2 11:23:51 PDT 2009
        - Make the search more robust to work around when FriendFeed search index it not up-to-date
        - Added -fi and -t to the usage doc

0.05  Wed Sep  2 02:18:55 PDT 2009
        - Added -f (force) and -t (test) command line switch

0.04  Wed Sep  2 01:18:08 PDT 2009
        - Rename the dist to App::CPAN::Fresh

0.03  Tue Sep  1 23:24:38 PDT 2009
        - If the found upload is older than 2 days, just use your mirror

0.02  Tue Sep  1 23:06:19 PDT 2009
        - Added -h option
        - Display usage information when there's no arguments

0.01  Tue Sep  1 22:03:51 2009
        - original version


lib/App/CPAN/Fast.pm
package App::CPAN::Fast;

use strict;
use 5.008_001;
our $VERSION = '0.03';

use base qw(App::Cmd::Simple);

use Carp;
use CPAN::Inject;
use Time::Piece;
use File::Temp;
use JSON;
use LWP::UserAgent;
use URI;

my $duration = 2 * 24 * 60 * 60;

sub opt_spec {
    return (
        [ "install|i", "install the module" ],
        [ "list|l", "list the recent uploads" ],
        [ "help|h", "displays usage info" ],
    );
}

sub execute {
    my($self, $opt, $args) = @_;

    if ($opt->{list} && $args->[0]) {
        $self->search($args->[0]);
    } elsif ($opt->{list}) {
        $self->recent;
    } elsif ($opt->{help} || !@$args) {
        $self->usage;
    } else {
        $self->install($args);
    }
}

sub recent {
    my $self = shift;
    my $res = $self->call("/feed/cpan");
    $self->display_results($res);
}

sub search {
    my($self, $q) = @_;
    my $res = $self->call("/search", { q => "$q group:cpan" });
    $self->display_results($res);
}

sub usage {
    require Pod::Usage;
    Pod::Usage::pod2usage(0);
}

sub install {
    my($self, $dists) = @_;

    my @install;
    for my $dist (@$dists) {
        my $path = $self->inject($dist);
        if ($path) {
            push @install, $path;
        } else {
            print "$dist is not found or not in the fresh uploads. Falling back to your mirror.\n";
            push @install, $dist;
        }
    }

    if (@install) {
        require CPAN;
        CPAN::Shell->install(@install);
    }
}

sub inject {
    my($self, $dist) = @_;
    $dist =~ s/::/-/g;

    my $res = $self->call("/search", { q => "$dist group:cpan" });
    for my $entry (@{$res->{entries}}) {
        my $info = $self->parse_entry($entry->{body}, $entry->{date}) or next;
        if ($info->{dist} eq $dist) {
            return $self->do_inject($info);
        }
    }

    return;
}

sub do_inject {
    my($self, $info) = @_;

    my $dir = File::Temp::tempdir(CLEANUP => 1);
    my $local = "$dir/$info->{dist}-$info->{version}.tar.gz";

    my $res = $self->new_ua->mirror($info->{url}, $local);
    if ($res->is_error) {
        croak "Fetching $info->{url} failed: ", $res->status_line;
    }

    CPAN::Inject->from_cpan_config->add(file => $local);
}

sub display_results {
    my($self, $res) = @_;
    for my $entry (@{$res->{entries}}) {
        my $info = $self->parse_entry($entry->{body}, $entry->{date}) or next;
        printf "%s-%s (%s)\n", $info->{dist}, $info->{version}, $info->{author};
    }
}

sub parse_entry {
    my($self, $body, $date) = @_;

    my $time = Time::Piece->strptime($date, "%Y-%m-%dT%H:%M:%SZ") or return;
    if (time - $time->epoch > $duration) {
        # entry found, but it's old
        return;
    }

    if ($body =~ /^([\w\-]+) ([0-9\._]*) by (.+?) - <a.*href="(http:.*?\.tar\.gz)"/) {
        return {
            dist    => $1,
            version => $2,
            author  => $3,
            url     => $4,
        };
    }

    return;
}

sub new_ua {
    my $self = shift;
    LWP::UserAgent->new(agent => "cpanf/$VERSION", env_proxy => 1);
}

sub call {
    my($self, $method, $opts) = @_;

    my $uri = URI->new("http://friendfeed-api.com/v2$method");
    $uri->query_form(%$opts) if $opts;

    my $ua  = $self->new_ua;
    my $res = $ua->get($uri);

    if ($res->is_error) {
        croak "HTTP error: ", $res->status_line;
    }

    JSON::decode_json($res->content);
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

App::CPAN::Fast - backend for I<cpanf> command

=head1 DESCRIPTION

App::CPAN::Fast is a backend for I<cpanf> command.

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<cpanf>

=cut


lib/App/CPAN/Fresh.pm
package App::CPAN::Fresh;

use strict;
use 5.008_001;
our $VERSION = '0.12';

use base qw(App::Cmd::Simple);

use Carp;
use Time::Piece;
use File::Temp;
use JSON;
use LWP::UserAgent;
use URI;

my $duration = 2 * 24 * 60 * 60;

sub opt_spec {
    return (
        [ "install|i", "install the module" ],
        [ "list|l", "list the recent uploads" ],
        [ "test|t", "test the dist" ],
        [ "force|f", "force install" ],
        [ "devel|d", "install even if it's a devel release" ],
        [ "minus|m", "use cpanminus" ],
        [ "help|h", "displays usage info" ],
    );
}

sub execute {
    my($self, $opt, $args) = @_;

    if ($opt->{list} && $args->[0]) {
        $self->search($args->[0]);
    } elsif ($opt->{list}) {
        $self->recent;
    } elsif ($opt->{help} || !@$args) {
        $self->usage;
    } else {
        $self->handle($opt, $args);
    }
}

sub recent {
    my $self = shift;
    my $res = $self->call("/feed/cpan");
    $self->display_results($res);
}

sub search {
    my($self, $q) = @_;
    my $res = $self->call("/search", { q => "$q group:cpan" });
    $self->display_results($res);
}

sub usage {
    require Pod::Usage;
    Pod::Usage::pod2usage(0);
}

sub handle {
    my($self, $opt, $dists) = @_;

    my @install;
    for my $dist (@$dists) {
        my $path = $self->inject($dist, $opt);
        if ($path) {
            push @install, $path;
        } else {
            print "$dist is not found or not in the fresh uploads. Falling back to your mirror.\n";
            push @install, $dist;
        }
    }

    my $method = "install";
    $method = "test" if $opt->{test};

    if (@install) {
        if ($opt->{minus}) {
            system "cpanm", ($opt->{force} ? "-f" : ()), @install;
        } else {
            require CPAN;
            if ($opt->{force}) {
                CPAN::Shell->force($method, @install);
            } else {
                CPAN::Shell->$method(@install);
            }
        }
    }
}

sub inject {
    my($self, $dist, $opt) = @_;
    $dist =~ s/::/-/g;

    for my $method ([ "/feed/cpan" ], [ "/search", { q => "$dist group:cpan" } ]) {
        my $res = $self->call($method->[0], $method->[1]);
        for my $entry (@{$res->{entries}}) {
            my $info = $self->parse_entry($entry->{body}, $entry->{date}) or next;
            if ($info->{dist} eq $dist) {
                if ($info->{version} =~ /_|-TRIAL/ && !$opt->{devel}) {
                    warn "$info->{dist}-$info->{version} found: No -d option, skipping\n";
                    return;
                }
                return $self->do_inject($info, $opt);
            }
        }
    }

    return;
}

sub do_inject {
    my($self, $info, $opt) = @_;

    if ($opt->{minus}) {
        return $info->{url};
    }

    my $dir = File::Temp::tempdir(CLEANUP => 1);
    my $local = "$dir/$info->{dist}-$info->{version}.tar.gz";

    print "Fetching $info->{url}\n";
    my $res = $self->new_ua->mirror($info->{url}, $local);
    if ($res->is_error) {
        croak "Fetching $info->{url} failed: ", $res->status_line;
    }

    require CPAN::Inject;
    CPAN::Inject->from_cpan_config->add(file => $local);
}

sub display_results {
    my($self, $res) = @_;
    for my $entry (@{$res->{entries}}) {
        my $info = $self->parse_entry($entry->{body}, $entry->{date}) or next;
        printf "%s-%s (%s)\n", $info->{dist}, $info->{version}, $info->{author};
    }
}

sub parse_entry {
    my($self, $body, $date) = @_;

    my $time = Time::Piece->strptime($date, "%Y-%m-%dT%H:%M:%SZ") or return;
    if (time - $time->epoch > $duration) {
        # entry found, but it's old
        return;
    }

    if ($body =~ /^([\w\-]+) (v?[0-9\._]*(?:-TRIAL)?) by (.+?) - <a.*href="(http:.*?\.tar\.gz)"/) {
        return {
            dist    => $1,
            version => $2,
            author  => $3,
            url     => $4,
        };
    }

    return;
}

sub new_ua {
    my $self = shift;
    LWP::UserAgent->new(agent => "cpanf/$VERSION", env_proxy => 1);
}

sub call {
    my($self, $method, $opts) = @_;

    my $uri = URI->new("http://friendfeed-api.com/v2$method");
    $uri->query_form(%$opts) if $opts;

    my $ua  = $self->new_ua;
    my $res = $ua->get($uri);

    if ($res->is_error) {
        croak "HTTP error: ", $res->status_line;
    }

    JSON::decode_json($res->content);
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

App::CPAN::Fresh - Query and install CPAN modules realtime from the fresh mirror

=head1 DEPRECATION

B<This module does not work anymore since FriendFeed has been shutdown.> You can use C<cpanm>'s C<--metacpan> option to enable MetaCPAN API search which can give you recently updated module in almost realtime. There's also C<--dev> option which allows you to search for development releases.

=head1 DESCRIPTION

App::CPAN::Fresh is a backend for I<cpanf> command.

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<cpanf>

=cut
